opt: panic with aggregates

Query:

```
SELECT array_agg(generate_series(1, 2))
```

Result:
```
panic: top-level relational expression cannot have outer columns: (1) [recovered]
        panic: panic while executing 1 statements: SELECT array_agg(generate_series(_, _)); caused by top-level relational expression cannot have outer columns: (1)

goroutine 138 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc425564a00, 0x6a67140, 0xc425450bc0, 0x60d3440, 0xc424f3e5a0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:675 +0x36f
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc425564a00, 0x6a67140, 0xc425450bc0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:414 +0x61
panic(0x60d3440, 0xc424f3e5a0)
        /usr/local/Cellar/go/1.10.2/libexec/src/runtime/panic.go:502 +0x229
github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*Optimizer).Optimize(0xc425565980, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:182 +0x200
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makeOptimizerPlan(0xc425564e30, 0x6a67200, 0xc424f21e00, 0x6a6a900, 0xc42ad26b80, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:460 +0x972
github.com/cockroachdb/cockroach/pkg/sql.(*planner).optionallyUseOptimizer(0xc425564e30, 0x6a67200, 0xc424f21e00, 0xc42573d02a, 0xf, 0xc42039d5fc, 0x4, 0x0, 0x1, 0x0, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/planner.go:537 +0xce
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc425564a00, 0x6a67200, 0xc424f21e00, 0x6a6a900, 0xc42ad26b80, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:759 +0x182
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc425564a00, 0x6a67200, 0xc424f21e00, 0x6a6a900, 0xc42ad26b80, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:403 +0xb42
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc425564a00, 0x6a67200, 0xc424f21e00, 0x6a6a900, 0xc42ad26b80, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:95 +0x341
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc425564a00, 0x6a67140, 0xc425450bc0, 0xc42034c400, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1079 +0x21a5
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc4207c1300, 0x6a67140, 0xc425450bc0, 0x0, 0x0, 0xc42573d067, 0x4, 0xc42573d02a, 0xf, 0x6a487c0, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:416 +0x1bb
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func3(0xc4207c1300, 0x6a67140, 0xc425450bc0, 0xc425742000, 0x5400, 0x15000, 0xc42095e830, 0xc42034c400, 0xc42034c3f0, 0xc42039d590, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:268 +0x122
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:267 +0xf1f
```

--- a/pkg/sql/logictest/testdata/logic_test/aggregate
+++ b/pkg/sql/logictest/testdata/logic_test/aggregate
@@ -1724,3 +1724,9 @@ query T
 SELECT string_agg('foo', CAST ((SELECT NULL) AS BYTES)) OVER ();
 ----
 foo
+
+# Regression test for #30166.
+query T
+SELECT array_agg(generate_series(1, 2))
+----
+{1,2}

--- a/pkg/sql/opt/optbuilder/groupby.go
+++ b/pkg/sql/opt/optbuilder/groupby.go
@@ -232,6 +232,9 @@ func (b *Builder) buildAggregation(
 	b.buildProjectionList(fromScope, projectionsScope)
 	b.buildOrderBy(fromScope, projectionsScope, orderByScope)
 	b.buildDistinctOnArgs(fromScope, projectionsScope, distinctOnScope)
+	if len(fromScope.srfs) > 0 {
+		fromScope.group = b.constructProjectSet(fromScope.group, fromScope.srfs)
+	}
 
 	aggInfos := aggOutScope.groupby.aggs
 

--- a/pkg/sql/opt/optbuilder/scope.go
+++ b/pkg/sql/opt/optbuilder/scope.go
@@ -868,7 +868,7 @@ func (s *scope) replaceAggregate(f *tree.FuncExpr, def *tree.FunctionDefinition)
 	defer s.builder.semaCtx.Properties.Restore(s.builder.semaCtx.Properties)
 
 	s.builder.semaCtx.Properties.Require(s.context,
-		tree.RejectNestedAggregates|tree.RejectWindowApplications|tree.RejectGenerators)
+		tree.RejectNestedAggregates|tree.RejectWindowApplications)
 
 	expr := f.Walk(s)
 	typedFunc, err := tree.TypeCheck(expr, s.builder.semaCtx, types.Any)

--- a/pkg/sql/opt/optbuilder/select.go
+++ b/pkg/sql/opt/optbuilder/select.go
@@ -450,11 +450,10 @@ func (b *Builder) buildSelectClause(
 		b.buildProjectionList(fromScope, projectionsScope)
 		b.buildOrderBy(fromScope, projectionsScope, orderByScope)
 		b.buildDistinctOnArgs(fromScope, projectionsScope, distinctOnScope)
+		if len(fromScope.srfs) > 0 {
+			fromScope.group = b.constructProjectSet(fromScope.group, fromScope.srfs)
+		}
 		outScope = fromScope
-	}
-
-	if len(fromScope.srfs) > 0 {
-		outScope.group = b.constructProjectSet(outScope.group, fromScope.srfs)
 	}
 
 	// Construct the projection.

--- a/pkg/sql/opt/optbuilder/testdata/aggregate
+++ b/pkg/sql/opt/optbuilder/testdata/aggregate
@@ -2853,3 +2853,23 @@ scalar-group-by
  └── aggregations
       └── array-agg [type=int[]]
            └── variable: y [type=int]
+
+# Regression test for #30166.
+build
+SELECT array_agg(generate_series(1, 2))
+----
+scalar-group-by
+ ├── columns: array_agg:2(int[])
+ ├── inner-join-apply
+ │    ├── columns: generate_series:1(int)
+ │    ├── values
+ │    │    └── tuple [type=tuple]
+ │    ├── zip
+ │    │    ├── columns: generate_series:1(int)
+ │    │    └── function: generate_series [type=int]
+ │    │         ├── const: 1 [type=int]
+ │    │         └── const: 2 [type=int]
+ │    └── true [type=bool]
+ └── aggregations
+      └── array-agg [type=int[]]
+           └── variable: generate_series [type=int]

